NewRegistration POCO Important FIX and Greater Flexibility for FileVaultProfile

mInternauta opened the pull request:
NewRegistration 
Extremely Important FIX! The LetsEncrypt API at this stage does not accept CONTACT as field but CONTACTS (in plural) will be accepted

FileVaultProfile
Optional Fix , Allows VaultProfile to be a full path and not relative to the current working path.

ebekker commented:
@mInternauta, your change to pluralize the Contacts field would break the client from interacting with the LE servers, the singular `Contact` is correct as per the [ACME protocol spec, section 5.2](https://ietf-wg-acme.github.io/acme/#rfc.section.5.2).

Indeed, the client code has been working fine in this regard for quite some time -- are you having some issue that motivated you to submit this property rename?

ebekker commented:
Additionally, the change to support absolute paths would not have any effect.  The `Path.Combine()` routine already does a [similar check internally](https://msdn.microsoft.com/en-us/library/fyy7a5kt(v=vs.110).aspx) to check if the second path argument is *rooted*.  If so, then the result of this call will just return the second path argument.

Here are some samples to illustrate:
```c#
> var p1 = @"foo\bar";
> var p2 = @"\abc\def";
> var p3 = @"c:\xzy";
> 
> System.IO.Path.Combine(p3, p1)
    "c:\\xzy\\foo\\bar"
> 
> System.IO.Path.Combine(p3, p2)
    "\\abc\\def"
> System.IO.Path.Combine(p2, p1)
    "\\abc\\def\\foo\\bar"
> System.IO.Path.Combine(p2, p3)
    "c:\\xzy"
```

ebekker commented:
However, I have some good news if you're looking for Vault to be rooted via an absolute path.  Branch [`cleanup-provider`](https://github.com/ebekker/ACMESharp/tree/cleanup-providers) is a major refactoring of the providers and extensions model in ACMESharp and this includes the way that Vault works.

With the introduction of *profiles*, the root directory of the Vault will now be defined using an absolute path and the current working directory will no longer have any impact on the resolved location of the Vault storage.  This branch will soon be ready for merging to `main` after I complete a few more tests and finish up the IIS provider, pby in the next few days.

mInternauta commented:
About checking the path in FileVaultProfile is just a way to work with Path.Combine in theory when it spent two full paths always the last is considered. Who guarantees at nearby versions of .NET stays working in that way.. But it's just an observation ..

"Why use something if it is not necessary to use it? (Design Pattern)"

But, on the Contact field. I changed because it seems to me the current LetsEncrypt  API dont follow the specification. Because I was facing the following message:
"Contact method is not supported," it accepts only the field as Contacts.

Thanks for the attention, and really god job with the project.

ebekker commented:
The error you got regarding contacts means you supplied a contact URL they don't understand.  Trying sending a mailto: URL as per the example in the spec.

By changing the field to plural, it meant the server was ignoring the contacts you sent.

ebekker commented:
If you start second-guessing each API call that its behavior and contract may change accidentally in a subsequent release, then you'll have to implement every piece of functionality yourself and you won't be able to rely on any of the BCLs or any third-party libraries.

You have to trust that everyone does their part, and if you find otherwise, submit a bug.  If they break their behavior contract, a lot of other projects, bigger and more important than our little ACME client here I'm sure, would scream faster and louder :-)

mInternauta commented:
Oops, I think I made myself half arrogant. You're right on the server to ignore the field. Okay I understand. But in fact that may lack a clear message from the server problem. And not a message saying that the field is not supported, but the email I informed is invalid right?

And the fact that you draw this conclusion is because it has much more knowledge on the API than me.

Anyway, it is my solution on a problem I faced. It's up to you to analyze and actually say whether they are valid or not. If they are not. So just close the Pull Request.

IYour project proved so good that I I'm using it as library for my little client. That only encapsulates your code and will make it into a fully automated Windows Service. You should be proud I live on the other side of the world (Brazil); (kkkk =p)

I appreciate your project. Opened a solution to a problem in Windows with certificates.

And I trust everyone =D Including you friend.

ebekker commented:
You are correct about the server error message, ACMESharp can use some improvement in this area to clean up capturing and propagating server error message details up.

I saw that you [commented](https://github.com/ebekker/ACMESharp/issues/13#issuecomment-170567249) that you're already working on implementing the PKI CSR generation using BouncyCastle for your own internals.

I think that ACMESharp would benefit greatly from a BouncyCastle implementation of the PKI Manager -- this would give it a more independent implementation, as well as prevent it from relying on external dependencies like OpenSSL or Windows CertEnroll.  Since you're already familiar with BC, let me know if you have any interest in implementing a BC-based PKI implementation.  If it works well, more than likely it would become the "default" internal implementation.

mInternauta commented:
I Performed a superficial implementation of BC, now i can generate a certificate with SAN Extension. However i still implementing the full signature of CertificateProvider.

I'm working on, but I can share some of BouncyCastle code. And how its generated the keys and certificate. I am involved in other projects but , i sure can share part of the code to help implement in the ACMESharp.

Another suggestion is to create a class that is a AcmeClient and implements all functions of POSH design in one location. (As the WebClient). So for those who want to use the ACMESharp within the .NET is much simpler: Take a look at AcmeFactory

It was also added support for SAN Extension, because the domains listed in the SAN property has to go through the authorization process in the ACME server. But it's just an example, you should have better way to implement this.

AcmeFactory: (https://github.com/mInternauta/IISAcmeClient/blob/master/src/IISAcmeClient.Core/AcmeFactory.cs)

Another suggestion, use the Windows Service, in my project I implemented a internal task scheduler  (Quartz.Net) it will automatically renew the certificates. So it is possible that ACMESharp also have an easy installation and configuration design, for those who dont understand programming but want to use the LetsEncrypt in Windows

ebekker commented:
If you would like to share whatever you've learned in working with BC, that would be great.  If it's already accessible via a GitHub project, you can just reference the project/code location.  If not and you don't mind posting it, you can just put it in a [gist](https://gist.github.com/) and either I or someone else will be able to reference the code for future work in ACMESharp, thanks!

Regarding your AcmeClient idea, there actually is already a project at [Lone-Coder/letsencrypt-win](https://github.com/Lone-Coder/letsencrypt-win) that uses the ACMESharp library, but wraps it inside of a stand-alone client that is more purpose-oriented for getting up and running quickly on Windows with IIS.  There's also a project at http://webprofusion.com/apps/certify which is a GUI client -- I'm not sure how far along that is, since it doesn't appear to be OSS, but in some of the discussions with the [author](https://github.com/ebekker/ACMESharp/issues/10#issuecomment-156808749), he mentions he's actually creating a PS session under the hood and calling into ACMESharp cmdlets.

Finally, a goal I've had from early on is similar to your idea -- it's to create a type of *master* PS cmdlet, `Lets-Encrypt` which actually breaks the conventions that are used in PS (decomposition), but mimics the usability aims of the official [letsencrypt client](https://github.com/letsencrypt/letsencrypt).  So my plan is to try to make the CLI interface (options, conventions, etc) of this master cmdlet resemble the official client as much as possible, and under the hood it just translates all of its calls into the other official ACMESharp PS module cmdlets.

Anyway, just a few thoughts...  Thanks for any contributions, and good luck.

mInternauta commented:
Okay, great ideias, thank you, You can look at: https://github.com/mInternauta/IISAcmeClient/blob/master/src/IISAcmeClient.Core/Acme/PKI/BouncyCertificateProvider.cs 
For the Bouncy code.. And good luck.. =D
